### ISSUE TITLE
How can I distinguish between types of failures with Result?

### ISSUE THREAD

**Issue by macfarmw** [vkhorikov/CSharpFunctionalExtensions#15]
I am using the ResultExtensions in a web API project. No matter what type of error happens in the railway oriented series of calls I am returning 400 (Bad Request). This is convenient and sufficient for now. Here is an excerpt of the code:

```
[Route("api/qrs/asbuilt/serialnumberstructure")]
[HttpPost]
public IHttpActionResult Post([FromBody] StructureRequestDto structureRequestDto)
{
    return
        ValidateRequest(structureRequestDto)
            .OnSuccess(request => AddNextSerialNumberIfMissing(request))
            .OnSuccess(request => AddPreviousSerialNumberIfMissing(request))
            .OnSuccess(request => AddStructureIfMissing(request))
            .OnSuccess(request => GetStructure(request))
            .Ensure(response => response.HasValue, FormatStructureNotFoundError(structureRequestDto))
            .Map(response => response.Value)
            .OnBoth(
                    response =>
                    response.IsSuccess ? 
                        this.Ok(response.Value) as IHttpActionResult :
                        this.BadRequest(response.Error) as IHttpActionResult);
}
```

I would like to improve the accuracy of my responses and return 500 for system errors and 400 when the error is caused by the input data. Because the error is string I'm not sure of a good way to do this. I thought about putting the HTTP status codes into the error strings and then parse at the end. That might work.

Are there any suggestions that would help me to be more specific with my error responses?

Thanks

**Comment by vkhorikov** [comment]
I would recommend putting a single generic handler to the very top of your execution stack to handle all 500s. The situation here is similar to exceptions. 5xx errors in web services are basically unexpected exceptions in the "exceptions terminology". Essentially, all 5xx errors mean bugs in your system because they are caused by the server itself, the user has nothing to do with them.

The moment you start doing anything in your code to handle that particular error/unexpected situation, it becomes excepted, and that means you can do something about it. That could be an actual operation in your system or a proper error message saying what user did wrong. Think of it as of "fixing" a bug, where you convert an unexpected situation into an excepted one and close that potential reason of failure.

The corollary from this is that you shouldn't return 500 errors intentionally, only 400s.

So, in short: 
5xx == unexpected situation/bug, a single handler for all of them should be enough
4xx == excepted situation, handle each of them separately in controllers

**Comment by macfarmw** [comment]
OK. Now I get it. I was trying to handle expected and unexpected errors with Result.Fail so it was really making the code messy. I have added the global error handler that logs the error details to Elmah and sends a generic error with HTTP status 500. The version of web API I am using (2.2) has a nice facility for global errors that was easy to setup. Perfect solution!

Thanks for your help and for the great library.
